File: main2.py
# Define the cycle costs for different operations
CYCLE_COST = {
    '+': 1,
    '-': 1,
    '*': 2,
    'Store': 3,  # Store operation cycle cost (simulated)
    'Load': 3    # Load operation cycle cost (simulated)
}

class Instruction:
    def __init__(self, dest, left_operand, operator, right_operand, cycle_cost):
        self.dest = dest
        self.left_operand = left_operand
        self.operator = operator
        self.right_operand = right_operand
        self.cycle_cost = cycle_cost

    @classmethod
    def parse(cls, line):
        # Split the line into two parts: before and after the equals sign
        parts = line.split(' = ')
        dest = parts[0].strip()
        expression = parts[1].strip()

        # Check if the operator is Store or Load (these are special instructions)
        if expression == 'Store' or expression == 'Load':
            # These are special instructions, where the operator is either Store or Load
            return cls(dest, None, expression, None, CYCLE_COST[expression])

        # Handle arithmetic instructions (must contain left operand, operator, right operand)
        expr = expression.split()
        if len(expr) != 3:
            raise ValueError(f"Invalid arithmetic instruction format: {line}")

        left_operand, operator, right_operand = expr
        cycle_cost = CYCLE_COST[operator.strip()]
        return cls(dest.strip(), left_operand.strip(), operator.strip(), right_operand.strip(), cycle_cost)

    def __str__(self):
        # Return the string representation of the instruction
        if self.operator in ['Store', 'Load']:
            return f"{self.dest} = {self.operator}"
        return f"{self.dest} = {self.left_operand} {self.operator} {self.right_operand}"

# Function to load and parse instructions from a file
def load_instructions(file_path):
    # Load instructions from a file and parse them into Instruction objects
    with open(file_path, 'r') as f:
        instructions = [Instruction.parse(line.strip()) for line in f if line.strip()]
    return instructions

class Scheduler:
    def __init__(self):
        self.busy_until = {}  # Tracks when each destination will be free

    def is_ready(self, instruction, current_cycle):
        # Check if dependencies (left_operand and right_operand) are resolved
        if instruction.left_operand and instruction.left_operand in self.busy_until:
            if current_cycle < self.busy_until[instruction.left_operand]:
                return False
        if instruction.right_operand and instruction.right_operand in self.busy_until:
            if current_cycle < self.busy_until[instruction.right_operand]:
                return False
        return True

    def reserve(self, instruction, current_cycle):
        # Reserve the destination until the instruction's cycle cost completes
        self.busy_until[instruction.dest] = current_cycle + instruction.cycle_cost


class Processor:
    def __init__(self, instructions):
        self.instructions = instructions
        self.scheduler = Scheduler()
        self.current_cycle = 0
        self.retired = set()  # Tracks indices of retired instructions
        self.in_flight = None  # Currently issued instruction
        self.wait_for_retire = False  # Flag to prevent issuing in the same cycle as retiring

    def run(self):
        instruction_index = 0
        total_instructions = len(self.instructions)

        # Print the header
        print(f"{'Cycle':<10}{'Issued Instruction':<30}{'Retired Instruction':<20}")
        print("-" * 60)

        while instruction_index < total_instructions or self.in_flight:
            self.current_cycle += 1

            issued_instruction = ""
            retired_instruction = ""

            # Retire the instruction if its execution is completed
            if self.in_flight:
                dest = self.in_flight.dest
                if self.scheduler.busy_until[dest] == self.current_cycle:
                    retired_index = self.instructions.index(self.in_flight) + 1
                    retired_instruction = f"Instruction {retired_index}"
                    self.retired.add(retired_index)
                    self.in_flight = None  # Clear the in-flight instruction
                    self.wait_for_retire = True  # Ensure next cycle does not issue immediately

            # Issue the next instruction if no instruction is in flight and no recent retirement
            if not self.in_flight and not self.wait_for_retire and instruction_index < total_instructions:
                next_instruction = self.instructions[instruction_index]
                if self.scheduler.is_ready(next_instruction, self.current_cycle):
                    issued_instruction = str(next_instruction)
                    self.scheduler.reserve(next_instruction, self.current_cycle)
                    self.in_flight = next_instruction
                    instruction_index += 1

            # Reset the wait flag after a cycle with no issuing
            if self.wait_for_retire and not issued_instruction:
                self.wait_for_retire = False

            # Print the current cycle's details in columns
            print(f"{self.current_cycle:<10}{issued_instruction:<30}{retired_instruction:<20}")

        print("-" * 60)
        print("Execution completed.")


# Load instructions and run the processor
instructions = load_instructions("instructions.txt")
processor = Processor(instructions)
processor.run()

File: main2.cs
/*
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
*/

//This code only has a first version of superscalar
//Is not working as intended
//It seems it is not staling the code when it needs to

public class Instruction
{
    public string Dest { get; }
    public string LeftOperand { get; }
    public string Operator { get; }
    public string RightOperand { get; }
    public int CycleCost { get; }

    private static readonly Dictionary<string, int> CycleCostEnum = new Dictionary<string, int>
    {
        { "+", 1 },
        { "-", 1 },
        { "*", 2 },
        { "Store", 3 },
        { "Load", 3 }
    };

    public Instruction(string dest, string leftOperand, string op, string rightOperand, int cycleCost)
    {
        Dest = dest;
        LeftOperand = leftOperand;
        Operator = op;
        RightOperand = rightOperand;
        CycleCost = cycleCost;
    }

    public static Instruction Parse(string line)
    {
        var parts = line.Split(" = ", StringSplitOptions.TrimEntries);
        if (parts.Length != 2)
            throw new ArgumentException($"Invalid instruction format: {line}");

        var dest = parts[0];
        var expression = parts[1];

        if (expression == "Store" || expression == "Load")
        {
            return new Instruction(dest, null, expression, null, CycleCostEnum[expression]);
        }

        var exprParts = expression.Split(' ');
        if (exprParts.Length != 3)
            throw new ArgumentException($"Invalid arithmetic instruction format: {line}");

        var leftOperand = exprParts[0];
        var op = exprParts[1];
        var rightOperand = exprParts[2];

        if (!CycleCostEnum.ContainsKey(op))
            throw new ArgumentException($"Invalid operator: {op}");

        return new Instruction(dest, leftOperand, op, rightOperand, CycleCostEnum[op]);
    }

    public override string ToString()
    {
        return Operator == "Store" || Operator == "Load"
            ? $"{Dest} = {Operator}"
            : $"{Dest} = {LeftOperand} {Operator} {RightOperand}";
    }
}

public class Scheduler
{
    private readonly Dictionary<string, int> _busyUntil = new();

    public bool IsReady(Instruction instruction, int currentCycle)
    {
        if (!string.IsNullOrEmpty(instruction.LeftOperand) && _busyUntil.TryGetValue(instruction.LeftOperand, out var leftBusyUntil))
        {
            if (currentCycle < leftBusyUntil) return false;
        }

        if (!string.IsNullOrEmpty(instruction.RightOperand) && _busyUntil.TryGetValue(instruction.RightOperand, out var rightBusyUntil))
        {
            if (currentCycle < rightBusyUntil) return false;
        }

        return true;
    }

    public void Reserve(Instruction instruction, int currentCycle)
    {
        _busyUntil[instruction.Dest] = currentCycle + instruction.CycleCost;
    }
}

public class Processor
{
    private readonly List<Instruction> _instructions;
    private readonly Scheduler _scheduler = new();
    private int _currentCycle;
    private HashSet<int> _retired = new();
    //private Instruction _inFlight;
    private bool _waitForRetire;

    private int _issue_slots;

    public Processor(List<Instruction> instructions, int Configuration, int issue_slots)
    {
        _instructions = instructions;
        _issue_slots = issue_slots;

        // Helps me run the correct configuration depending on what was asked when creating the class
        if (Configuration == 1)
        {
            Single_Instruction_run();

        }
        else if (Configuration == 2)
        {

            Superscalar_in_order_run();

        }
        else if (Configuration == 3)
        {

            Superscalar_out_of_order_run();

        }

    }

    public void Single_Instruction_run()
    {
        int instructionIndex = 0;
        int totalInstructions = _instructions.Count;
        Instruction currentInstruction = null;
        int retireCycle = 0;
        bool waitForNextCycle = false; // Flag to track waiting after retirement

        Console.WriteLine($"{"Cycle",-10}{"Issued Instruction",-30}{"Retired Instruction",-20}");
        Console.WriteLine(new string('-', 60));

        while (instructionIndex < totalInstructions || currentInstruction != null)
        {
            _currentCycle++;
            string issuedInstruction = "";
            string retiredInstruction = "";

            // Retire the current instruction if its cycle cost is completed
            if (currentInstruction != null && _currentCycle == retireCycle)
            {
                int retiredIndex = _instructions.IndexOf(currentInstruction) + 1;
                retiredInstruction = $"Instruction {retiredIndex}";
                currentInstruction = null;
                waitForNextCycle = true; // Set the flag to wait for the next cycle
            }

            // Issue a new instruction only if:
            // 1. There is no current instruction in-flight.
            // 2. We're not waiting for the next cycle after a retirement.
            if (currentInstruction == null && !waitForNextCycle && instructionIndex < totalInstructions)
            {
                currentInstruction = _instructions[instructionIndex];
                issuedInstruction = currentInstruction.ToString();
                retireCycle = _currentCycle + currentInstruction.CycleCost;
                instructionIndex++;
            }

            // Reset the wait flag if it was set during the previous cycle
            if (waitForNextCycle && currentInstruction == null)
            {
                waitForNextCycle = false;
            }

            Console.WriteLine($"{_currentCycle,-10}{issuedInstruction,-30}{retiredInstruction,-20}");
        }

        Console.WriteLine(new string('-', 60));
        Console.WriteLine("Execution completed.");
    }

    private void Superscalar_in_order_run()
    {
        int totalInstructions = _instructions.Count;
        int instructionIndex = 0;
        var inFlightInstructions = new List<(Instruction instruction, int retireCycle)>();

        Console.WriteLine($"{"Cycle",-10}{"Issued Instructions",-30}{"Retired Instructions",-20}");
        Console.WriteLine(new string('-', 60));

        while (instructionIndex < totalInstructions || inFlightInstructions.Any())
        {
            _currentCycle++;
            var issuedInstructions = new List<string>();
            var retiredInstructions = new List<string>();

            // Retire instructions that have completed
            for (int i = inFlightInstructions.Count - 1; i >= 0; i--)
            {
                var (instruction, retireCycle) = inFlightInstructions[i];
                if (_currentCycle >= retireCycle)
                {
                    int retiredIndex = _instructions.IndexOf(instruction) + 1;
                    retiredInstructions.Add($"Instruction {retiredIndex}");
                    inFlightInstructions.RemoveAt(i);
                }
            }

            // Issue instructions up to the issue slot limit
            while (inFlightInstructions.Count < _issue_slots && instructionIndex < totalInstructions)
            {
                var instruction = _instructions[instructionIndex];
                if (_scheduler.IsReady(instruction, _currentCycle))
                {
                    issuedInstructions.Add(instruction.ToString());
                    _scheduler.Reserve(instruction, _currentCycle);
                    inFlightInstructions.Add((instruction, _currentCycle + instruction.CycleCost));
                    instructionIndex++;
                }
                else
                {
                    break; // Dependencies are not resolved; stop issuing
                }
            }

            Console.WriteLine($"{_currentCycle,-10}{string.Join(", ", issuedInstructions),-30}{string.Join(", ", retiredInstructions),-20}");
        }

        Console.WriteLine(new string('-', 60));
        Console.WriteLine("Execution completed.");
    }

    private void Superscalar_out_of_order_run()
    {
        int totalInstructions = _instructions.Count;
        var readyQueue = new Queue<Instruction>();
        var inFlightInstructions = new List<(Instruction instruction, int retireCycle)>();
        var issuedIndices = new HashSet<int>();

        Console.WriteLine($"{"Cycle",-10}{"Issued Instructions",-30}{"Retired Instructions",-20}");
        Console.WriteLine(new string('-', 60));

        while (issuedIndices.Count < totalInstructions || inFlightInstructions.Any())
        {
            _currentCycle++;
            var issuedInstructions = new List<string>();
            var retiredInstructions = new List<string>();

            // Retire instructions that have completed
            for (int i = inFlightInstructions.Count - 1; i >= 0; i--)
            {
                var (instruction, retireCycle) = inFlightInstructions[i];
                if (_currentCycle >= retireCycle)
                {
                    int retiredIndex = _instructions.IndexOf(instruction) + 1;
                    retiredInstructions.Add($"Instruction {retiredIndex}");
                    inFlightInstructions.RemoveAt(i);
                }
            }

            // Populate ready queue with all ready instructions
            for (int i = 0; i < totalInstructions; i++)
            {
                if (!issuedIndices.Contains(i) && _scheduler.IsReady(_instructions[i], _currentCycle))
                {
                    readyQueue.Enqueue(_instructions[i]);
                    issuedIndices.Add(i); // Mark instruction as queued
                }
            }

            // Issue instructions from the ready queue up to the issue slot limit
            while (readyQueue.Count > 0 && inFlightInstructions.Count < _issue_slots)
            {
                var instruction = readyQueue.Dequeue();
                issuedInstructions.Add(instruction.ToString());
                _scheduler.Reserve(instruction, _currentCycle);
                inFlightInstructions.Add((instruction, _currentCycle + instruction.CycleCost));
            }

            Console.WriteLine($"{_currentCycle,-10}{string.Join(", ", issuedInstructions),-30}{string.Join(", ", retiredInstructions),-20}");
        }

        Console.WriteLine(new string('-', 60));
        Console.WriteLine("Execution completed.");
    }

}

public static class Program
{
    public static void Main()
    {

        //Needs to make sure the file path of the instructions.txt is the correct one, this cannot be shortened 

        var instructions = LoadInstructions("C:\\Users\\jairo\\OneDrive\\Documentos\\GitHub\\Computer-Architecture-Project\\instructions.txt");
        //Small test that makes sure the instruction are read in the correct way
        /*
        foreach (var instruction in instructions)
        {
            // Print the class name of each instruction
            Console.WriteLine(instruction);
        }
        */

        //Runs the processor with sinlge instruction, in order, with one issue slot
        var processor = new Processor(instructions, 1, 1);
        var processor2 = new Processor(instructions, 2, 1);
        var processor3 = new Processor(instructions, 3, 1);
    }

    private static List<Instruction> LoadInstructions(string filePath)
    {
        return File.ReadAllLines(filePath)
                   .Where(line => !string.IsNullOrWhiteSpace(line))
                   .Select(Instruction.Parse)
                   .ToList();
    }
}